gtk/gtkprintoperation-portal.c

```diff
@@ -0,0 +1,597 @@
+
+  if (!gxdp_print_call_print_finish (portal->proxy, &handle, result, &error))
+    {
+      g_warning ("Error: %s", error->message);
+      g_error_free (error);
+      g_main_loop_quit (portal->loop);
+      return;
+    }
+
+  portal->print_response_signal_id =
+    g_dbus_connection_signal_subscribe (g_dbus_proxy_get_connection (G_DBUS_PROXY (portal->proxy)),
+                                        "org.freedesktop.portal.Desktop",
+                                        "org.freedesktop.portal.Request",
+                                        "Response",
+                                        handle,
+                                        NULL,
+                                        G_DBUS_SIGNAL_FLAGS_NO_MATCH_RULE,
+                                        print_response,
+                                        portal, NULL);
+
+  g_free (handle);
+}
+
+GtkPrintOperationResult
+gtk_print_operation_portal_run_dialog (GtkPrintOperation *op,
+                                       gboolean           show_dialog,
+                                       GtkWindow         *parent,
+                                       gboolean          *do_print)
+{
+  GtkPrintOperationPrivate *priv = op->priv;
+  GVariant *settings;
+  GVariant *setup;
+  GVariantBuilder opt_builder;
+  GVariant *options;
+  guint signal_id;
+  PortalData *portal;
+  GtkPrintOperationResult result;
+
+  portal = g_new0 (PortalData, 1);
+  portal->proxy = get_portal_proxy ();
+  portal->op = g_object_ref (op);
+  portal->parent = parent;
+  portal->result = GTK_PRINT_OPERATION_RESULT_CANCEL;
+  portal->print_cb = NULL;
+  portal->destroy = NULL;
+
+  signal_id = g_signal_lookup ("create-custom-widget", GTK_TYPE_PRINT_OPERATION);
+  if (g_signal_has_handler_pending (op, signal_id, 0, TRUE))
+    g_warning ("GtkPrintOperation::create-custom-widget not supported with portal");
+
+  g_variant_builder_init (&opt_builder, G_VARIANT_TYPE_VARDICT);
+  options = g_variant_builder_end (&opt_builder);
+
+  if (priv->print_settings)
+    settings = gtk_print_settings_to_gvariant (priv->print_settings);
+  else
+    {
+      GVariantBuilder builder;
+      g_variant_builder_init (&builder, G_VARIANT_TYPE_VARDICT);
+      settings = g_variant_builder_end (&builder);
+    }
+
+  if (priv->default_page_setup)
+    setup = gtk_page_setup_to_gvariant (priv->default_page_setup);
+  else
+    {
+      GtkPageSetup *page_setup = gtk_page_setup_new ();
+      setup = gtk_page_setup_to_gvariant (page_setup);
+      g_object_unref (page_setup);
+    }
+
+  portal->loop = g_main_loop_new (NULL, FALSE);
+
+  gxdp_print_call_print (portal->proxy,
+                         "",
+                         "Print Test",
+                         settings,
+                         setup,
+                         options,
+                         NULL,
+                         print_called,
+                         portal);
+
+  gdk_threads_leave ();
+  g_main_loop_run (portal->loop);
```

Comment:
This is a pretty much straight copy from what we do in gtkprintoperation-unix.c.
Also, I'm not entirely sure that calling the gdbus sync() methods from inside gtk is actually safe wrt to gdk locking. As you can see here, we are dropping the gdk lock before entering a recursive mainloop. the sync() call will also enter a recursive mainloop, but it won't drop the gdk lock.